accord-core/src/main/java/accord/coordinate/FetchData.java

```diff
@@ -415,14 +444,92 @@ public Void apply(SafeCommandStore safeStore)
+        }
+
+        // if can only propagate Truncated, we might be stale; try to upgrade the
+        private Known applyOrUpgradeTruncated(SafeCommandStore safeStore, SafeCommand safeCommand, Command command)
+        {
+            // if our peers have truncated this command, then either:
+            // 1) we have already applied it locally; 2) the command doesn't apply locally; 3) we are stale; or 4) the command is invalidated
+            if (command.saveStatus().isUninitialised())
+                return null; // TODO (expected): maybe stale?
+
+            if (command.hasBeen(PreApplied))
+            {
+                updateDurability(safeStore, safeCommand);
+                return null;
+            }
+
+            if (Infer.safeToCleanup(safeStore, command, route, full.executeAt))
+            {
+                Commands.setErased(safeStore, safeCommand);
+                return null;
+            }
+
+            // we're now at least partially stale, but let's first see if we can progress this shard, or we can do so in part
+            Timestamp executeAt = command.executeAtIfKnown(full.executeAtIfKnown());
+            if (executeAt == null)
+            {
+                // we don't even know the execution time, so we cannot possibly proceed besides erasing the command state and marking ourselves stale
+                // TODO (required): we could in principle be stale for future epochs we haven't witnessed yet. Ensure up to date epochs before finalising this application.
+                safeStore.commandStore().markShardStale(safeStore, txnId, route.participants().toRanges(), false);
+                Commands.setErased(safeStore, safeCommand);
+                return null;
+            }
+
+            Ranges executeRanges = safeStore.ranges().allBetween(txnId, executeAt);
+            executeRanges = safeStore.commandStore().redundantBefore().expectToExecute(txnId, executeAt, executeRanges);
+
+            if (executeRanges.isEmpty())
+            {
+                // TODO (expected): we might prefer to adopt Redundant status, and permit ourselves to later accept the result of the execution and/or definition
+                Commands.setTruncatedApply(safeStore, safeCommand, route);
+                return null;
+            }
+
+            Known required = PreApplied.minKnown;
+            Known requireExtra = required.subtract(command.known());
+            Ranges achieveRanges = full.sufficientFor(requireExtra, executeRanges);
+            Participants<?> participants = route.participants().slice(executeRanges, Minimal);
+
+            if (participants.isEmpty())
+            {
+                // we only coordinate this transaction, so being unable to retrieve its state does not imply any staleness
+                Commands.setTruncatedApply(safeStore, safeCommand, route);
+                return null;
+            }
+
+            Ranges staleRanges = executeRanges.subtract(achieveRanges);
```

Comment:
It might be that we want to impose some limit on the number of unique ranges we permit to be stale before we begin merging them. But equally we don't want to mark much larger ranges as stale than necessary. 

But yes, correct - we cannot find enough information to progress the transaction's execution, and we know it should be there. So we're now behind and cannot expect to catch up.

Perhaps we should attempt to contact more than a majority in this case, before marking stale. But this is probably a later refinement - I can add a TODO (and a comment)